### ISSUE TITLE
Image result not correctly in Mac retina display

### ISSUE THREAD

**Issue by iamkwan** [simov/screenshot-capture#1]
Hello,
When i try to crop some area using Macbook pro retina
the result of image not match with my cropped target
maybe in retina display, the pixel is very high, so the result of image cannot display correctly

**Comment by simov** [comment]
I don't own a monitor with such high resolution/dpi, nor do I have a Mac, so I probably won't be able to reproduce it on my end.

You can at least do the following:
1. Switch to *Crop and Wait* mode
2. Select some area for cropping and take a screenshot of your entire window with external app
3. Save the selected area for cropping with the extension
4. Post both screenshots here

**Comment by iamkwan** [comment]
entire window of cropping some area
[https://upload.cc/i3/oAuYaw.png](https://upload.cc/i3/oAuYaw.png)

result of image
[https://upload.cc/i/C4Qzb6.png](https://upload.cc/i/C4Qzb6.png)

**Comment by simov** [comment]
It looks like there is a way to detect the retina display, which is good. I'll try to reproduce it on my end, and I'll let you know when I have something ready to test.

**Comment by iamkwan** [comment]
also when save the selected area, the file of image missing the extension
here is my solution
./content/content.js

`function save (image) {
  document.location.href = image.replace('image/png', 'image/octet-stream')
}`

replace to

`function save (image) {
  var link = document.createElement('a');
  link.download = "download.png";
  link.href = image.replace('image/png', 'image/octet-stream');
  link.click();
}`

**Comment by simov** [comment]
I'll try that out, this is on my todo list too :+1:

**Comment by simov** [comment]
BTW the code responsible for cropping the image is located in background.js - [drawImage](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage). My best bet is that the crop area passed from the web page should be scaled there. Then the only thing left would be to detect the higher dpi inside the web page and pass it to the background page along with the cropping information.

**Comment by iamkwan** [comment]
agree, i find someone got this problem too in stackoverflow
[http://stackoverflow.com/questions/32013884/chrome-extension-screenshot-partial-image-cropping-for-retina-display](http://stackoverflow.com/questions/32013884/chrome-extension-screenshot-partial-image-cropping-for-retina-display)

**Comment by simov** [comment]
@iamkwan can you try something for me?

I am assuming that you already know how to clone this repo and the extensions as *unpacked* one.

Can you modify this code in background.js:

```js
context.drawImage(img,
  left * 2, top * 2,
  width * 2, height * 2,
  0, 0,
  width * 2, height * 2
)
```

And then reload the extension and crop something.

`2` here should be the same as executing `window.devicePixelRatio` in your browser console, but it's most probably `2`. I'm yet to find a way to reliably reproduce the dpi stuff on my end, so if that works for you I'll push a fix and ask you to test again.

If anything fails post screenshots here. Thanks!

**Comment by iamkwan** [comment]
still not match but improved

entire window of cropping some area
[https://upload.cc/i3/Es1hqY.png](https://upload.cc/i3/Es1hqY.png)

result of image
[https://upload.cc/i/Kh8Esi.png](https://upload.cc/i/Kh8Esi.png)

**Comment by simov** [comment]
The top left corner is in the correct place, but it looks like the width and the height should be multiplied by `4` Definitely weird, I'll let you know if I come up with something else.

**Comment by iamkwan** [comment]
also when I try to crop in jcrop demo page
the position of crop is wrong, maybe the jcrop demo page have some class or id same with this extension?
I have a idea, when click the extension, capture the current screen and send the data image to new tab to let user crop

**Comment by simov** [comment]
@iamkwan can you pull the [fix-dpi](https://github.com/simov/screenshot-capture/tree/fix-dpi) branch and test, I think I finally got it.

**Comment by iamkwan** [comment]
@simov the problem was solve, but i think i will change the resolution of the cropped image to for retina to normal
I have one question, do you know how to insert some text center in the screen? for example when user click the extension icon to trigger the jcrop, then insert the text in the screen without draw canvas (eg. Select Area), when user onmousedown, the text will hide.

**Comment by simov** [comment]
> but i think i will change the resolution of the cropped image from retina to normal

I think in this case the image will look smaller on your high dpi monitor. I should probably add this as an option in the options page.

For your second question: It should be pretty easy to add such message to the web page. 

The best way to indicate that the extension is ready for cropping is to dim the page a little bit, like it's done in the jcrop demo, but the problem is that this approach doesn't play well with the extension. Simply because there is no background image to set opacity on. Trust me I've spent a good amount of time on that and there is no apparent way to do it.

**Comment by iamkwan** [comment]
Not smaller, is the quality lower in my retina display when I change from
retina to normal
In the fix-pixel extension, when I crop the image
Eg. 300x300
But the result is retina image, so the width and high of result image is
more than 300x300

**Comment by simov** [comment]
Yep, I'm going to add this as an option to this branch. I'll let you know when it's ready to test it.

**Comment by simov** [comment]
I also like your idea about using some other sort of indication when the extension is ready to crop. I'll think about it but this will most probably be implemented in the next version, after I publish the DPI fix and the option to preserve the size of the cropped area.

**Comment by iamkwan** [comment]
also when i try to insert element (eg. button) into screen when jcrop onSelect, but when i try to click the element, it trigger jcrop, cannot trigger the element onclick, even i set this element to z-index: 999999999

**Comment by simov** [comment]
It depends on where you insert the element :)

**Comment by iamkwan** [comment]
@simov LOL my mistake, it can trigger my function now

**Comment by iamkwan** [comment]
but the onChange is let me hard to make other function, when user just onmousedown, it will trigger the jcrop onChange event............

**Comment by iamkwan** [comment]
@simov do you know how to detect when user move the crop box without using onChange?

**Comment by simov** [comment]
I don't think there is a default event for that (changing the top/left position of a DOM element).

**Comment by iamkwan** [comment]
finally i use this if condition inside onChange even to detect user is cropping image not just click the area of the screen, because when user just click the crop box it will trigger onChange too
`if(oldx != e.x || oldx2 != e.x2 || oldy != e.y || oldy2 != e.y2)`

**Comment by simov** [comment]
I just pushed the new DPI option to the [fix-dpi](https://github.com/simov/screenshot-capture/tree/fix-dpi) branch. Let me know what do you think.

**Comment by iamkwan** [comment]
i think the options is good, but you need to let user know they can edit this in option (let they know the extension have option to change some config

**Comment by simov** [comment]
Well, they have to find the options page, there is no other way :) I have outlined how to get to the options page in the extension's description in the chrome store.

What's left is to implement the same option for the *Capture Viewport* option and then I'm publishing the fix.

Open up another issue for the capture indication overlay and the file name issue if you want.

**Comment by iamkwan** [comment]
The file name is fixed using my solution

**Comment by simov** [comment]
Ok, I'll take a look at it later on.

**Comment by simov** [comment]
Just pushed the fix for the *Capture Viewport* method to be aware of the new DPI option.

**Comment by iamkwan** [comment]
@simov do you know chrome extension is start to only for chrome OS?
I'm just furious.
[https://blog.chromium.org/2016/08/from-chrome-apps-to-web.html](https://blog.chromium.org/2016/08/from-chrome-apps-to-web.html)

**Comment by simov** [comment]
_Chrome **Apps**_ not _Chrome **Extensions**_

**Comment by iamkwan** [comment]
oh, thank you @simov , you saved my life

**Comment by simov** [comment]
btw I just merged the DPI fix to master and added the file name fix, take a look at it if you want

**Comment by iamkwan** [comment]
it works, but i found some problem, i will open a new issue for this problem

**Comment by simov** [comment]
Ok, I'm closing this one. Thanks for the feedback :+1: I'll think about the overlay too.

**Comment by simov** [comment]
The fixes are published with verison 1.4
